### ISSUE TITLE
Integrate turican0's changes

### ISSUE THREAD

**Issue by GrimSqueaker** [thobbsinteractive/magic-carpet-2-hd#73]
I think it makes sense to integrate the changes from @turican0's development branch.
One huge advantage is that the 64 bit build from there is playable which is important for me since it simplifies a lot of things.
I have taken a quick look and tried to merge the branches and after that to rebase the development branch from @thobbsinteractive on @turican0's development branch. This seems to be a lot of work - especially for me, as I am not familiar with all the changes.
Do you have any idea how we could achieve this?

**Comment by turican0** [comment]
Hi, there were a lot of minor changes and fixes in my code. I was comparing the differences with the original pixel by pixel and byte by byte, so it would be good to move it carefully, otherwise the errors I fixed will come back. I don't know how many changes there are since the last merge, but it might be easier to use my branch as a base and apply the changes to it, your branch, that have happened since the last version.

I also worked on multiplayer, but didn't finish it. A lot of bugs are fixed, but it would take , someone who understands the network more to look at it. I tried, but I couldn't emulate the original logic on the current network libraries :)

**Comment by thobbsinteractive** [comment]
I didn't know you were still working on it Tom. That's good news.
Yes I think merging to my Dev branch is a good idea, but ideally only
complete features and change by change. Cherry picking night be the best
option here.

I am stunned to hear that multiplayer is being worked on. Perhaps we can
leave it on a branch until it is working? It will also make it easier for
someone to come in later and complete it.

Last time I merged the changes it took almost week to do. I am worried
about all the refactoring I have done, moving code out of Sub_Main, it will
make things difficult. But if we succeed, we can finally merge my fork
changes back to the original repo.

**Comment by turican0** [comment]
Hey, I'm not working on it anymore. Last time was a year ago. I was working on multiplayer at the time.
I put most of the network code in port_net.cpp. But it doesn't really work. It should emulate NETBIOS/NETBEUI via the modern network library(boost/asio.hpp).
In nebios, all computers register their names, and then they can send messages to each other. I've tried making a stack that runs in a second process that stores all incoming messages. From which the game process can pick them off one by one. But maybe I went about it the wrong way :)
I tried not to change the original code too much(just fix the reverse engineering bugs), so that I don't have to debug the logic of the original code. But even in
I also fixed the display of multiplayer menus(character selection) and multiplayer levels.

**Comment by GrimSqueaker** [comment]
Hi,
so, I had a look at it yesterday and I think the most feasible way is to take HEAD of turican0/development and merge into it thobbsinteractive/development. This results in only 20-30 conflicting files and the only non-trivial conflicts are in sub_main.cpp.
And also there it should not be too difficult to figure out which code got moved into other cpp files and port the changes.

I will have a go at this, but would kindly ask the two of you for a review afterwards or some help if I get stuck.

**Comment by thobbsinteractive** [comment]
Sounds good Seb.

I have a better knowledge of the code now, so this should be easier for me
to review this time.

The new render works now (for all level types), so we can kill the old one
if it is still in there. I just hope git hub recognises where I have moved
some structures and the render code out into their own files or this could
be very difficult.

Good luck!

**Comment by turican0** [comment]
Hey, watch the rendering. My rendering was tweaked frame by frame exactly according to the original - at the beginning of all levels, in the middle, etc. So it should be with minimal errors.
Before removing the old rendering, you would want to compare pixel by pixel to see if they match. Including shadows, reflections, dragons with "3D" circles necks, etc. Maybe you've done similar tests, and I'm afraid unnecessarily :)
But the debugging took a lot of work and it would be a shame to lose it.

Anyway, after removing the old rendering and moving the functions from Sub_Main. you can no longer compare the game's behavior with the original DOS version byte by byte, so hopefully some unsolvable bug won't appear.

**Comment by GrimSqueaker** [comment]
@turican0 , could you please briefly describe your workflow to guarantee 100% picture frame-by-frame compatibility?
Or did you already write this down last year? Sorry, I can't remember.

In remc2 we might be even able to automate this as kind of an regression test. 
For this, I think, the debugging defines would need to be transformed into flags that can be controlled via command line interface (e.g. load a in-level-savegame, provide fixed input, render a few frames, compare to the same frames from the original).
But I do not know how to get frames from the original MC2 - especially in an automated way.

@thobbsinteractive , is it possible to have both implementations at the same time and switching between them? I mean the original rendering and your improvements? This way we would not have to keep two repositories in sync.

On the other hand: Is 100% rendering compatibility to the original something that should be kept up forever? One could also make the point that "being close enough" and providing other improvements is also a value on its own.
Just brainstorming here, but if we can answer that question then it would clarify how to work with the repositories in the future.

**Comment by thobbsinteractive** [comment]
The new code in single threaded mode is basically Toms render method that
included the ability to set the resolution above 640. I believe that you
should get the same results in 640x480 (I am not sure about the shadows)

I would also also love to know how a frame by frame comparison was done and
see the results.

I personally want to keep the software render, it's a fascinating bit of
history. I believe I have fixed all the rendering bugs, cave levels work,
the shading error in reflections was fixed last week and all the menus and
buttons align (except the help text). I want to get it running properly in
1080 (if that is even possible with a software render!), but the ultimate
goal is to implement OpenGL along side it, that is why I refactored the
Render code into a separate class. So that one day people can switch
between the two in the settings and we can use some advanced OpenGL
effects. Maybe even have the option (like the Doom modes) to have 3d models
for the sprites and RTX support.

**Comment by turican0** [comment]
Hi, I've described this before, but I'll try again.

1) To get the data you need a modified dosbox in my dosbox-x-remc2 repository.

2)It does a few things, but the main thing is that there is an enginestep() procedure in the engine.cpp file that runs with each instruction. Using
writeseqall(0x237B05);, certain structures are stored each time an address is passed through. for example, in the writeseqall procedure you see:
writesequence(adress, 0x10000, 320 * 200, 0x3aa0a4, skip);
screen state.
If you respect dosbox on a different machine, the address prefixes may be different(but not necessarily), you need to work them out.

3)Already in the new game engine I placed the following code, which compared the recorded data with the real ones:
#ifdef DEBUG_SEQUENCES
	add_compare(0x237B05, debugafterload);
#endif //DEBUG_SEQUENCES

**Comment by turican0** [comment]
This data contains the states of terrain maps, creatures and screens. So any deviation means an error in the code. (Except for one level, where pseudo-random deviations cannot be replicated.)
If you can handle regression tests - load saves from some levels and compare data. I can record this data from those places.

**Comment by turican0** [comment]
//#define TEST_NETWORK - this line in my branch is only for test network, for normal game must deactivate it

**Comment by GrimSqueaker** [comment]
> I fix linux32 workflow in my developement branch(my master branch is obsolete now) - I use Ubuntu 22.04, Ubuntu 20.04 have problems with some dependeces.

Nice, I didn't try that.

**Comment by thobbsinteractive** [comment]
Tonight I have a lot of other stuff to do, but with the bit of free time I have I am going to look into the issue HD crash. But I will gladly look to your branch tomorrow and Thursday.

**Comment by turican0** [comment]
I will try to integrate the changes into my repository gradually, but carefully in steps.
My repository(development) is now free of known bugs, and it's more important to have a game that can be played from start to finish than a cool graphic but untweaked code. Believe me, finishing the game and removing all possible new bugs is much harder than writing new code.
The ideal would be to keep the diagnostics option, and also keep the old and new rendering, which could be toggled with #ifdef.
Then you could always do regression tests to see if it's a new or old bug.
So I'll try to make changes to make the repos as close as possible and maybe I'll figure out other bugs.
Anyway, it would like to get to a stage where there is one default repository to push changes to and the others are just like playgrounds.

**Comment by thobbsinteractive** [comment]
I have played my branch from the first to the last level. I also fixed two
bugs that crashed the game in the middle of two different levels . If you
turn off the multi threading and set the res to 640 I believe you will have
the authentic experience.

Either way,  I think you will find a lot in there you like for instance
using SDL to scale the graphics. Maybe we even fixed some of the same bugs.

I agree that the repos should be compatible. I think the forks can be there
to add new functionality like OpenGL.

You can see a list of my plans in my forks read me, I imagine we have
similar goals.

One way or another we will get this synched up. All my changes are made
incrementally and usually for just one bug fix at a time, so merging them
should not be too difficult

**Comment by thobbsinteractive** [comment]
So I have an idea. I could put the old render back in and call it if 640 is selected without threading. That way @turican0 can still do the byte comparison.
Let me know if you think it is a good idea and I will get it done.

**Comment by turican0** [comment]
Hi, that's a good idea(for 320 or 640). In the meantime, I'm bringing my repository closer to yours. I prefer to do it detailed and slow, even if it is more laborious.
I've seen a lot of variables renamed to better names, but I've also seen bits of code where a few minor changes and fixes have changed them back. So I want to go through it change by change to l the best possible result. It will be a little while before I start posting any patches.
What will be the main repository for the changes?
integration/integrate_thobbsinteractive_into_turican0
or
thobbsinteractive/magic-carpet-2-hd?
(One of the main problems why merging is so difficult now is the splitting of repositories, unfortunately I can't accept Thobbs changes after commits now as I have a number of things changed before)

I want to run regression tests in steps during the merge.

Once the merge is done I'd like to focus on network deployment again. I bought an Intel NUC 11 miniPC to go with the TV, it only has 64GB but Visual Studio will fit in it. I can access it via remote desktop.  So I can easily control two instances of the game from one computer :)

**Comment by thobbsinteractive** [comment]
You can always switch to 320 with the R key even in the threaded render, I worked hard to keep that.

I am looks at integration/integrate_thobbsinteractive_into_turican0 already has all my changes. I just checked in a fix for a build error, but the render code crashes early in the first level.

@turican0 you might want to avoid that branch and create a refresh branch in my or @GrimSqueakers repo if you want to integrate changes step-by-step.

My Repo or @GrimSqueaker repo is easiest to integrate because it is forked from my repo, merging (once complete) is easy.

So there were 2 game breaking crashes I fixed.

One was an indexing error, so I made a minor change in index calculation to match the original code. 
The other bigger change was here:
https://github.com/thobbsinteractive/magic-carpet-2-hd/pull/53

I simply had to convert the code back because I could not figure out how to fix the new code. But the Level 13 worked and I could load and save it mid game too. If you can fix it using your original code that would be great.

Good luck.

**Comment by thobbsinteractive** [comment]
@GrimSqueaker integration/integrate_thobbsinteractive_into_turican0 is crashing with the reflections turned on :(. There's an issue somewhere. I will take a look at the changes tomorrow night

**Comment by turican0** [comment]
Thanks, okay, I'll base my aunts on your repository. And I'll be sending back minor patches, for files with minor changes, so that the differences will gradually decrease and the repositories will be unified.

I have a question, I noticed you added X and Y resolution to the Events.

GenerateEvents_49290(&D41A0_0.terrain_2FECE, 1, 640, 480);
I've passed it through to this function:
void SetMousePositionInMemory_5BDC0(int16_t posX, int16_t posY, uint16_t screenWidth)
But screenWidth is not used there.
Are there functions in Events that need resolution?
(ones where it was originally hardcoded)

Thanks for the answer, Tom.

**Comment by thobbsinteractive** [comment]
I believe it is called when moving the cursor to the spell or subcategory of spell you have selected.
It looks like  screenWidth is not used. I think I was changing a whole load of similar methods and included it accidentally. I am not usually that messy :)

**Comment by GrimSqueaker** [comment]
Hi,
sorry for being late to the discussion. I think my repo is **not** a good choice as it is more like a playground and your repos have the big improvements. I will use my fork only for working on branches and the open PRs against what ever is best for you.
That said, @thobbsinteractive, I think it is not worth your time to work on the integration/integrate_thobbsinteractive_into_turican0 branch any longer if Tom starts to integrate your improvements. I started this integration in the assumption that this is not going back to the @turican0 repo.

Hints:
The last common ancestor of your development branches is
```
> git merge-base integration/integrate_turrican0_updates  turrican0/development 
455944d92b667d4cc5aa8f992b917d723acb8ad1
> git log --pretty -n 1 455944d92b667d4cc5aa8f992b917d723acb8ad1
commit 455944d92b667d4cc5aa8f992b917d723acb8ad1
Author: turican0 <41465132+turican0@users.noreply.github.com>
Date:   Wed Mar 31 18:41:42 2021 +0200

    210331-03
```

As it seems there have been only 29 commits to thobbsinteractive/development between HEAD and the common ancestor 455944d92.
```
> git log --oneline --first-parent 455944d92b667d4cc5aa8f992b917d723acb8ad1..thobbs/development | wc -l
29
> git log --oneline --first-parent 455944d92b667d4cc5aa8f992b917d723acb8ad1..thobbs/development 
debded07f (thobbs/development) Merge pull request #72 from thobbsinteractive/refactor/remove-duplicate-thread-code
ac392d3bd (integration/integrate_turrican0_updates) Merge pull request #71 from thobbsinteractive/fix-reflection-render
f332ca597 (thdevelopment) Merge pull request #70 from thobbsinteractive/multi-threaded-render
b59fd8895 Merge pull request #67 from GrimSqueaker/integration/thdev
02b915e64 Merge pull request #66 from thobbsinteractive/fix-spell-selection
778d792d9 Merge pull request #64 from thobbsinteractive/re-implement-super-fast
2b8eb2dbb Merge pull request #63 from thobbsinteractive/add-development-guide
b09d79e14 Merge pull request #62 from thobbsinteractive/fix-mana-goal
7f78ffdd6 Merge pull request #58 from thobbsinteractive/implement-hardware-res-scaling
7152d1264 Merge pull request #56 from thobbsinteractive/refactor-level-selection-screen
486fa1372 Merge pull request #53 from thobbsinteractive/level-13-crash
2e7f3cace Merge pull request #49 from thobbsinteractive/level-9-crash-fix
728bf530c Merge pull request #48 from thobbsinteractive/multi-threaded-render
301957b48 Merge pull request #46 from thobbsinteractive/correctly-chat-menu-display-in-resolutions-above-640
6713323f3 Merge pull request #44 from thobbsinteractive/move-game-render-code
5fe608e02 Merge pull request #38 from thobbsinteractive/In-game-option-menu-text-is-missing-in-1920x1080
66525598e Update README.md
cecf9c16c Update README.md
830982e94 Update README.md
8d1b3be80 Merge pull request #36 from thobbsinteractive/Spiderwebs-hitting-players-are-in-left-corner
78e4d8609 Merge pull request #35 from thobbsinteractive/fix-health-bar-render
5c23c90ee Merge pull request #33 from thobbsinteractive/Cave-Maps-are-missing-half-the-Terrain-on-the-right-#30
8926412c7 Merge branch 'GrimSqueaker-development' into development
07dfa4b13 Merge pull request #29 from thobbsinteractive/merge-of-development-from-GrimSquearker-development
f14f9dab6 Merge pull request #28 from thobbsinteractive/incremental-merge-of-pull-request
694cb2990 Merge pull request #27 from thobbsinteractive/resolution-upgrade
f5514aa08 Merge pull request #26 from thobbsinteractive/resolution-upgrade
261e0d129 Update README.md
875d653cd Merge pull request #23 from GrimSqueaker/development
```
So, it might be easier to step along these logical feature merges instead of doing a single-step merge as I did on my integration branch.

**Comment by thobbsinteractive** [comment]
I agree. 
We can see how @turican0  gets on. I have been trying this evening myself to put the old render into GameRender.cpp. It is going okay, but I might wait to see how the merge goes.
Worse case, I can create a merge branch on my repo. That might actually be easiest solution. Although your work @GrimSqueaker in your merge branch has at least shown how much there is to merge. I think it is almost working. Perhaps we can leave it as a backup.

**Comment by turican0** [comment]
Hi, I've started merging our repositories. Unfortunately, there are so many small changes that it can't be done one commit at a time. I made a copy of your repository, showing the differences using Winmerge.
For each difference, I fix either my repository or yours, so that all the new stuff and fixes remain, but the files are the same at the end.
I have now unified part of the engine directory.
Once I unify the core files I will continue with extracting features in files like Type_D93C0_Bldgprmbuffer.cpp. sub_main.cpp I will save for last :)
I hope to be done in 1-2 weeks.
Please try not to make big changes. Alternatively, please confirm my commits before making them.

There were a lot of small fixes missing in your repository, I think it makes sense.

Fingers crossed, Tom.

**Comment by PieterVdc** [comment]
once there're closer wouldn't making a github organization and all working on the same master be better then each having a fork on your personal accounts?
personally not gonna do mc development so it's up to you guys but that just sounds like the most efficient way

**Comment by turican0** [comment]
We are basically trying to do something similar now - to have one main repository(now thobbsinteractive/magic-carpet-2-hd). The other repositories and branches would be just for testing new features and improvements. But first we need to unify the repositories.

**Comment by thobbsinteractive** [comment]
@GrimSqueaker, can we close this now?

**Comment by turican0** [comment]
Yes, I think the merge went fine, if we run into something we won't address it in this thread anymore.
